jobfunnel/config/cli.py

```diff
@@ -0,0 +1,408 @@
+        help='Reconstruct a new master CSV file from all available cache files.'
+             'WARNING: this will replace all the statuses/etc in your master '
+             'CSV, it is intended for starting fresh / recovering from a bad '
+             'state.'
+    )
+
+    parser.add_argument(
+        '--no-scrape',
+        action='store_true',
+        help='Do not make any get requests, and attempt to load from cache.'
+    )
+
+    # Proxy stuff
+    # TODO: subparser.
+    parser.add_argument(
+        '-protocol',
+        dest='proxy_protocol',
+        type=str,
+        help='Proxy protocol.'
+    )
+    parser.add_argument(
+        '-ip',
+        dest='proxy_ip',
+        type=str,
+        help='Proxy IP (V4) address.'
+    )
+    parser.add_argument(
+        '-port',
+        dest='proxy_port',
+        type=str,
+        help='Proxy port address.'
+    )
+
+    # Delay stuff
+    # TODO: move delay args into a subparser for improved -h clarity
+    parser.add_argument(
+        '--delay-random',
+        dest='delay_random',
+        action='store_true',
+        help='Turn on random delaying for certain get requests.'
+    )
+
+    parser.add_argument(
+        '--delay-converging',
+        dest='delay_converging',
+        action='store_true',
+        help='Use converging random delay for certain get requests.'
+    )
+
+    parser.add_argument(
+        '-delay-max',
+        dest='delay_max_duration',
+        type=float,
+        help='Set delay seconds for certain get requests.'
+    )
+
+    parser.add_argument(
+        '-delay-min',
+        dest='delay_min_duration',
+        type=float,
+        help='Set lower bound value for delay for certain get requests.'
+    )
+
+    parser.add_argument(
+        '-delay-algorithm',
+        choices=[a.name for a in DelayAlgorithm],
+        help='Select a function to calculate delay times with.'
+    )
+
+    return vars(parser.parse_args())
+
+
+def config_parser(args_dict: Dict[str, Any]) -> Dict[str, Any]:
+    """Parse the JobFunnel configuration settings and combine CLI, YAML and
+    defaults to build a valid config dictionary for initializing config objects.
+
+    FIXME: still doesn't handle CLI args only properly...
+    """
+    # Init and pop args that are cli-only and not in our schema
+    settings_yaml_file = args_dict.pop('settings_yaml_file')
+    output_folder = args_dict.pop('output_folder')
+    args_dict.pop('do_recovery_mode')  # NOTE: this is handled in __main__
+    config = {'search': {}, 'delay': {}, 'proxy': {}}
+
+    # NOTE: these are mutually exclusive from output_folder
+    user_passed_paths = bool(
+       args_dict['master_csv_file'] or args_dict['cache_folder']
+       or args_dict['block_list_file'] or args_dict['duplicates_list_file']
+    )
+
+    # Build a config that respects CLI, defaults and YAML
+    # NOTE: we a passed settings YAML first so we can inject CLI after if needed
+    if settings_yaml_file:
```

Comment:
If we decided to make the YAML file and CLI mutually exclusive, then could be a place where we could decide to make sure there aren't any CLI args being passed and get everything from YAML file.